xknx/core/value_reader.py

```diff
@@ -9,51 +9,64 @@
 """
 import asyncio
 import logging
+from typing import TYPE_CHECKING, Optional
 
-from xknx.telegram import Telegram, TelegramType
+from xknx.telegram import GroupAddress, Telegram, TelegramType
+
+if TYPE_CHECKING:
+    import xknx.XKNX as XKNX
 
 logger = logging.getLogger("xknx.log")
 
 
 class ValueReader:
     """Class for reading the value of a specific KNX group address from KNX bus."""
 
-    # pylint: disable=too-many-instance-attributes
-
-    def __init__(self, xknx, group_address, timeout_in_seconds=2):
+    def __init__(
+        self, xknx: "XKNX", group_address: GroupAddress, timeout_in_seconds: float = 2.0
+    ):
         """Initialize ValueReader class."""
         self.xknx = xknx
-        self.group_address = group_address
+        self.group_address: GroupAddress = group_address
         self.response_received_or_timeout = asyncio.Event()
-        self.success = False
-        self.timeout_in_seconds = timeout_in_seconds
-        self.timeout_handle = None
-        self.received_telegram = None
+        self.success: bool = False
+        self.timeout_in_seconds: float = timeout_in_seconds
+        self.received_telegram: Optional[Telegram] = None
 
-    async def read(self):
+    async def read(self) -> Optional[Telegram]:
         """Send group read and wait for response."""
         cb_obj = self.xknx.telegram_queue.register_telegram_received_cb(
             self.telegram_received
         )
-
         await self.send_group_read()
-        await self.start_timeout()
-        await self.response_received_or_timeout.wait()
-        await self.stop_timeout()
 
-        self.xknx.telegram_queue.unregister_telegram_received_cb(cb_obj)
+        try:
+            await asyncio.wait_for(
+                self.response_received_or_timeout.wait(),
+                timeout=self.timeout_in_seconds,
+            )
+        except asyncio.TimeoutError:
+            logger.warning(
+                "Error: KNX bus did not respond in time (%s secs) to GroupValueRead request for: %s",
+                self.timeout_in_seconds,
+                self.group_address,
+            )
+        finally:
+            # cleanup to not leave callbacks (for asyncio.CancelledError)
+            self.xknx.telegram_queue.unregister_telegram_received_cb(cb_obj)
```

Comment:
The core of this PR is always this pattern. 
Instead of using `loop.call_later(self.timeout_in_seconds, self.timeout)` in `self.start_timeout()` with `self.stop_timeout()` and `self.timeout()`
we use `asyncio.wait_for()` with a timeout argument and move the contents of `self.timeout()` into the exception clause so asyncio takes care of handling the timeout for us. This way we can use `finally` to always remove our callbacks.